File: nodejs-typescript/src/domain/Account.ts
import type { AccountNumber } from './AccountNumber.js';
import { InsufficientFundsError } from './errors.js';
import type { Money } from './Money.js';

export class Account {
  constructor(
    readonly accountNumber: AccountNumber,
    private balance: Money,
  ) {}

  getBalance(): Money {
    return this.balance;
  }

  credit(amount: Money): void {
    this.balance = this.balance.add(amount);
  }

  canDebit(amount: Money): boolean {
    return this.balance.canSubtract(amount);
  }

  debit(amount: Money): void {
    if (!this.canDebit(amount)) {
      throw new InsufficientFundsError('Insufficient funds');
    }
    this.balance = this.balance.subtract(amount);
  }

  replaceBalance(balance: Money): void {
    this.balance = balance;
  }
}

File: dotnet/src/MableBank.Domain/Account.cs
using MableBank.Domain.Errors;

namespace MableBank.Domain;

public sealed class Account
{
    private readonly ICreditPolicy _creditPolicy;

    public Account(
        AccountNumber accountNumber,
        Money balance,
        ICreditPolicy? creditPolicy = null)
    {
        AccountNumber = accountNumber;
        Balance = balance;
        _creditPolicy = creditPolicy ?? DefaultCreditPolicy.Instance;
    }

    public AccountNumber AccountNumber { get; }

    public Money Balance { get; private set; }

    public bool CanDebit(Money amount) => Balance.CanSubtract(amount);

    public void Credit(Money amount) =>
        Balance = _creditPolicy.ApplyCredit(Balance, amount);

    public void Debit(Money amount)
    {
        if (!CanDebit(amount))
        {
            throw new InsufficientFundsException("Insufficient funds");
        }

        Balance = Balance.Subtract(amount);
    }

    public void ReplaceBalance(Money balance) => Balance = balance;

    public Account Copy() => new(AccountNumber, Balance, _creditPolicy);
}
